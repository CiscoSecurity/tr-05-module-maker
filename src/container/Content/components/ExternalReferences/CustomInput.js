import styles from "./ExternalReferences.module.scss";
import React from "react";

export default class CustomInput extends React.Component {
    render() {
        return(
            <div className={styles.reference_wrapper}>
                <input type="text" placeholder="Enter label" className={styles.custom_input}
                       autoComplete="off"/>
                <input type="text" placeholder="Enter link" className={styles.custom_input}
                       autoComplete="off"/>
            </div>
        )
    }
}
