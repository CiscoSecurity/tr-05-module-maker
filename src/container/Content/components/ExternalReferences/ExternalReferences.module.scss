@import "src/globals/constants/colors";

.reference_wrapper {
    display: flex;
    align-items: center;
}
.predefined_input {
    flex: 1;
    margin: 4px;
}
.custom_input {
    flex: auto;
    margin: 4px;
}

.predefined_input {
    &:disabled {
        background-color: $input-disabled-background;
    }
    flex: 1;
    margin: 4px;
}
