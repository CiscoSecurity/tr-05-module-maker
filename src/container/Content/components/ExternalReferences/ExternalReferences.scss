@import "src/globals/constants/colors";

p > button {
  flex: 1;
  min-width: 75px;
  max-height: 28px;
  color: $white-color;
  background-color: $blue-background;
  border-radius: 3px;
  border: 0;
  box-shadow: none;
  font-weight: 400;
  line-height: 1.5;
  text-overflow: ellipsis;
  text-align: center;
  vertical-align: middle;
  cursor: pointer;
  white-space: nowrap;
  margin: 4px;

  &:hover {
    background-color: $hover-color;
  }
}
