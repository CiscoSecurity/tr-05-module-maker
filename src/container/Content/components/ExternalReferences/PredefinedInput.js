import styles from "./ExternalReferences.module.scss";
import React from "react";
import {toggleLink, updateLink} from "../../../../redux/actions";
import {connect} from "react-redux";


class PredefinedInput extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            link: ""
        }
    }

    changeCheckboxHandler = (event) => {
        this.props.toggleLink(event.target.name);
        this.setState(prev => ({
            ...prev, ...{
                link: ""
            }
        }))
    }

    changeInputHandler = (event) => {
        event.persist()
        this.setState(prev => ({
            ...prev, ...{
                link: event.target.value.trim()
            }
        }))
        this.props.updateLink(
            {
                label: event.target.name,
                link: event.target.value
            }
            );
    }

    render() {
        return (
            <div className={styles.reference_wrapper}>
                <input type="checkbox" autoComplete="off" name={this.props.label}
                       onChange={this.changeCheckboxHandler}/>
                {this.props.label}
                <input
                    type="text"
                    name={this.props.label}
                    autoComplete="off"
                    value={this.state.link}
                    className={styles.predefined_input}
                    onChange={this.changeInputHandler}
                    disabled={this.props.syncReferences.filter(el =>
                        Object.values(el).includes(this.props.label)
                    ).length === 0}
                />
            </div>
        )
    }
}

const mapDispatchToProps = {
    toggleLink,
    updateLink
}

const mapStateToProps = state => {
    return {
        syncReferences: state.json.external_references
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(PredefinedInput);
