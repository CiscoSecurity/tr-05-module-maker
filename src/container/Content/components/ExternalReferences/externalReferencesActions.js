import { ADD_EXTERNAL_REFERENCE, DELETE_EXTERNAL_REFERENCE, UPDATE_EXTERNAL_REFERENCE }
from "globals/constants/types";
import { v4 as uuidv4 } from 'uuid';


export const updateExternalReference = (id, pair) => ({
        type: UPDATE_EXTERNAL_REFERENCE,
        payload: {id: id, pair: pair}
})

export const addExternalReference = () => ({
        type: ADD_EXTERNAL_REFERENCE,
        payload: {
            id: uuidv4()
        }
})

export const deleteExternalReference = (id) => ({
        type: DELETE_EXTERNAL_REFERENCE,
        payload: id
})
