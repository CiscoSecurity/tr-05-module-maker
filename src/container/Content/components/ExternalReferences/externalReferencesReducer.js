import {
    ADD_EXTERNAL_REFERENCE,
    DELETE_EXTERNAL_REFERENCE,
    UPDATE_EXTERNAL_REFERENCE,
} from "globals/constants/types"

const initialState = []

export const externalReferencesReducer = (state = initialState, action) => {
    switch (action.type) {
        case ADD_EXTERNAL_REFERENCE:
            return state.concat(
                {
                    "label": "",
                    "link": "",
                    "id": action.payload.id
                }
            )

        case DELETE_EXTERNAL_REFERENCE:
            return state.filter(
                el => !Object.values(el).includes(action.payload)
            )

        case UPDATE_EXTERNAL_REFERENCE:
            return state.map(
                    item => {
                        if (item.id !== action.payload.id) {
                            return item
                        }
                        return {
                            ...item,
                            [action.payload.pair.name]: action.payload.pair.value
                        }
                    }
                )

        default:
            return state
    }
}
