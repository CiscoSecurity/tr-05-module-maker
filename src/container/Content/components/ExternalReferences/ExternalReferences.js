import React from "react";
import PredefinedInput from "./components/PredefinedInput/PredefinedInput";
import CustomInput from "./components/CustomInput/CustomInput";
import * as Constants from "globals/constants/constants";


class ExternalReferences extends React.Component {
    render() {
        return (
            <div>
                <p>External References</p>
                <fieldset>
                    {
                        Constants.LABELS.map(
                            item => (<PredefinedInput label={item}/>)
                        )
                    }
                    <CustomInput/>
                    <CustomInput/>
                </fieldset>
            </div>
        )
    }
}

export default ExternalReferences;
