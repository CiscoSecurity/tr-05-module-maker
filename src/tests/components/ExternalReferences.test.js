import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import ExternalReferences from "../../container/Content/components/ExternalReferences/ExternalReferences";
import React from "react";
import configureStore from 'redux-mock-store';
import toJson from "enzyme-to-json";
import * as mocks from "../mocks/mockPayloads"


const mockStore = configureStore([]);


describe('ExternalReferences', () => {
    it('should render ExternalReferences without CustomInputs', () => {
        const store = mockStore({
            external_references: [],
        });

        const wrapper = mount(
            <Provider store={store}>
                <ExternalReferences/>
            </Provider>);
        expect(wrapper.find('CustomInput')).toHaveLength(0);
        expect(toJson(wrapper)).toMatchSnapshot();
    })

    it('should render ExternalReferences with several CustomInputs', () => {
        const store = mockStore({
            external_references:
                mocks.EXTERNAL_REFERENCES
        });

        const wrapper = mount(
            <Provider store={store}>
                <ExternalReferences/>
            </Provider>);
        expect(wrapper.find('CustomInput')).toHaveLength(2);
        expect(toJson(wrapper)).toMatchSnapshot();
    })
});
