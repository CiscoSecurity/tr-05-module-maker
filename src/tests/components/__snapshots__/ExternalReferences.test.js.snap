// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExternalReferences should render ExternalReferences with several CustomInputs 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ExternalReferences)>
    <ExternalReferences
      addExternalReference={[Function]}
      syncExternalReferences={
        Array [
          Object {
            "id": "00000000-0000-0000-0000-000000000003",
            "label": "Sign Up",
            "link": "https://haveibeenpwned.com/API/Key",
          },
          Object {
            "id": "00000000-0000-0000-0000-000000000004",
            "label": "FAQs",
            "link": "https://haveibeenpwned.com/FAQs",
          },
        ]
      }
    >
      <div
        className="reference-section"
      >
        <div>
          <label
            className="input-label"
          >
             
            External References
            <button
              onClick={[Function]}
              type="button"
            >
              + Add
            </button>
          </label>
        </div>
        <Connect(CustomInput)
          key="00000000-0000-0000-0000-000000000003"
          reference={
            Object {
              "id": "00000000-0000-0000-0000-000000000003",
              "label": "Sign Up",
              "link": "https://haveibeenpwned.com/API/Key",
            }
          }
        >
          <CustomInput
            deleteExternalReference={[Function]}
            reference={
              Object {
                "id": "00000000-0000-0000-0000-000000000003",
                "label": "Sign Up",
                "link": "https://haveibeenpwned.com/API/Key",
              }
            }
            updateExternalReference={[Function]}
          >
            <div
              className="reference_wrapper"
            >
              <input
                autoComplete="off"
                className="custom_input"
                list="ext-ref"
                name="label"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="Enter label"
                required={true}
                type="text"
                value="Sign Up"
              />
              <input
                autoComplete="off"
                className="custom_input"
                name="link"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="Enter link"
                required={true}
                type="text"
                value="https://haveibeenpwned.com/API/Key"
              />
              <datalist
                id="ext-ref"
              >
                <option
                  key="Sign Up 00000000-0000-0000-0000-000000000003"
                >
                  Sign Up
                </option>
                <option
                  key="Free Trial 00000000-0000-0000-0000-000000000003"
                >
                  Free Trial
                </option>
                <option
                  key="Contact Us 00000000-0000-0000-0000-000000000003"
                >
                  Contact Us
                </option>
              </datalist>
              <svg
                className="closeIcon"
                onClick={[Function]}
              >
                <use
                  xlinkHref="sprite.svg#icon-small-x-close"
                />
              </svg>
            </div>
          </CustomInput>
        </Connect(CustomInput)>
        <Connect(CustomInput)
          key="00000000-0000-0000-0000-000000000004"
          reference={
            Object {
              "id": "00000000-0000-0000-0000-000000000004",
              "label": "FAQs",
              "link": "https://haveibeenpwned.com/FAQs",
            }
          }
        >
          <CustomInput
            deleteExternalReference={[Function]}
            reference={
              Object {
                "id": "00000000-0000-0000-0000-000000000004",
                "label": "FAQs",
                "link": "https://haveibeenpwned.com/FAQs",
              }
            }
            updateExternalReference={[Function]}
          >
            <div
              className="reference_wrapper"
            >
              <input
                autoComplete="off"
                className="custom_input"
                list="ext-ref"
                name="label"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="Enter label"
                required={true}
                type="text"
                value="FAQs"
              />
              <input
                autoComplete="off"
                className="custom_input"
                name="link"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="Enter link"
                required={true}
                type="text"
                value="https://haveibeenpwned.com/FAQs"
              />
              <datalist
                id="ext-ref"
              >
                <option
                  key="Sign Up 00000000-0000-0000-0000-000000000004"
                >
                  Sign Up
                </option>
                <option
                  key="Free Trial 00000000-0000-0000-0000-000000000004"
                >
                  Free Trial
                </option>
                <option
                  key="Contact Us 00000000-0000-0000-0000-000000000004"
                >
                  Contact Us
                </option>
              </datalist>
              <svg
                className="closeIcon"
                onClick={[Function]}
              >
                <use
                  xlinkHref="sprite.svg#icon-small-x-close"
                />
              </svg>
            </div>
          </CustomInput>
        </Connect(CustomInput)>
      </div>
    </ExternalReferences>
  </Connect(ExternalReferences)>
</Provider>
`;

exports[`ExternalReferences should render ExternalReferences without CustomInputs 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ExternalReferences)>
    <ExternalReferences
      addExternalReference={[Function]}
      syncExternalReferences={Array []}
    >
      <div
        className="reference-section"
      >
        <div>
          <label
            className="input-label"
          >
             
            External References
            <button
              onClick={[Function]}
              type="button"
            >
              + Add
            </button>
          </label>
        </div>
      </div>
    </ExternalReferences>
  </Connect(ExternalReferences)>
</Provider>
`;
